### ISSUE TITLE
Issue with Identifier Types and Composite Foreign Keys (kanel + kysely)

### ISSUE THREAD

**Issue by dcousineau** [kristiandupont/kanel#571]
**TLDR** Found possibly an edge case, possibly an issue on my end, involving composite foreign keys and would love guidance on next steps (fix my kanel configs? learn where in kanel this could be happening an help contribute a fix? wait for a fix?)

I have recently come across an issue where Kanel (+ Kanel-Kysely) generates problematic outputs as my schema has evolved to have more complex compound foreign keys. It's possible my configurations are bad or I'm doing something very silly.

> [!IMPORTANT]
> I have created a [code sandbox MVP reproduction](https://codesandbox.io/p/devbox/condescending-sky-h4748g) to illustrate what happens. This sandbox has my exact `.kanelrc.js` along with the schema below.

The TLDR of the issue is when I have a moderately complex n:m type join table in situations where both sides of the join have composite primary keys. Kanel attempts to do an `OR` union of the "identifier" types (but for only one side of the equation) for identifier types it... doesn't actually create.

Given the following postgres schema:

```pgsql
CREATE TABLE organization (
  id bigserial PRIMARY KEY
);
CREATE TABLE a (
  id bigint NOT NULL,
  organization_id bigint NOT NULL REFERENCES organization(id),
  CONSTRAINT a_pkey PRIMARY KEY (id, organization_id)
);
CREATE TABLE b (
  id bigint NOT NULL,
  organization_id bigint NOT NULL REFERENCES organization(id),
  CONSTRAINT b_pkey PRIMARY KEY (id, organization_id)
);
CREATE TABLE a_b_match (
  organization_id bigint NOT NULL,
  a_id bigint NOT NULL,
  b_id bigint NOT NULL,
  CONSTRAINT a_b_match_a_fkey FOREIGN KEY (organization_id, a_id) REFERENCES a(organization_id, id),
  CONSTRAINT a_b_match_b_fkey FOREIGN KEY (organization_id, b_id) REFERENCES b(organization_id, id),
  CONSTRAINT a_b_match_pkey PRIMARY KEY (organization_id, a_id, b_id)
);
```

And a `.kanelrc.js` almost entirely based on https://github.com/kristiandupont/kanel/blob/main/example/.kanelrc.js, we get the following outputs.

> [!NOTE]
> Specifically notice how `ABMatch.ts` is attempting to import `BOrganizationId` from `B.ts` and union it with `OrganizationId`, _however_ `B.ts` doesn't actually define or export `BOrganizationId`

In an ideal world, Kanel would **not** attempt to create `BOrganizationId` nor would it attempt to union it with `OrganizationId`.

#### `ABMatch.ts`
```typescript
// @generated
// This file is automatically generated by Kanel. Do not modify manually.

import { type OrganizationId } from './Organization';
import { type BOrganizationId, type BId } from './B';
import { type AId } from './A';
import { type ColumnType, type Selectable, type Insertable, type Updateable } from 'kysely';

/** Represents the table public.a_b_match */
export default interface ABMatchTable {
  organization_id: ColumnType<OrganizationId | BOrganizationId, OrganizationId | BOrganizationId, OrganizationId | BOrganizationId>;

  a_id: ColumnType<AId, AId, AId>;

  b_id: ColumnType<BId, BId, BId>;
}

export type ABMatch = Selectable<ABMatchTable>;

export type NewABMatch = Insertable<ABMatchTable>;

export type ABMatchUpdate = Updateable<ABMatchTable>;
```

#### `B.ts`

```typescript
// @generated
// This file is automatically generated by Kanel. Do not modify manually.

import { type OrganizationId } from './Organization';
import { type ColumnType, type Selectable, type Insertable, type Updateable } from 'kysely';

/** Identifier type for public.b */
export type BId = string;

/** Represents the table public.b */
export default interface BTable {
  id: ColumnType<BId, BId, BId>;

  organization_id: ColumnType<OrganizationId, OrganizationId, OrganizationId>;
}

export type B = Selectable<BTable>;

export type NewB = Insertable<BTable>;

export type BUpdate = Updateable<BTable>;
```

**Comment by kristiandupont** [comment]
Hmm, yes, I see the problem but I admit that I am not sure where in the chain the bug lies. 
As with many other issues, I won't promise when I can get around to debugging it but I *would* like to fix this as it does appear to be a bug and not some out-of-scope feature request. If this is blocking you, a post-render hook is probably the easiest way to get around it

**Comment by dcousineau** [comment]
Sounds great. Except for me managing to stub my toes here kanel has been excellent to me.

I'll try the post render hook first just to bandaid the situation. If you have any examples handy (that aren't already in this repo) that would be helpful, but if not no worries!

I'll also keep you posted if I am able to debug or chase the issue down myself.

**Comment by dcousineau** [comment]
Something I just noticed because I bumped to `3.9.0` 2 weeks ago (but hadn't had cause to re-run type generation) is I am now getting a lot of "Could not resolve circular reference" warnings (from this line: https://github.com/kristiandupont/kanel/blob/ace86e45955b2eca64b0f66a2a367b9a3cc50a30/packages/kanel/src/generators/resolveType.ts#L100) which was from #540 and the following error:

```
TypeError: Cannot read properties of undefined (reading 'join')
    at writeFile (./node_modules/kanel/build/writeFile.js:41:27)
    at ./node_modules/kanel/build/processDatabase.js:78:60
    at Array.forEach (<anonymous>)
    at processDatabase (./node_modules/kanel/build/processDatabase.js:78:18)
    at async main (./node_modules/kanel/build/cli/main.js:120:9
```

(Upon further inspection the `writeFile` method has a `lines` argument set to undefined rather than an array)

So this is a clue at least, I can't spend much more time today digging in but I'll report back when I get the chance to spend more time on it.

**Comment by stayradiated** [comment]
I've just started re-factoring a database schema to use composite foreign keys and I am also hitting a similar issue.

The SQL change is:

```diff
ALTER TABLE ONLY public.document
     ADD CONSTRAINT "document:foreignKey(owned_by_person_id,person.id)"
-    FOREIGN KEY (owned_by_person_id) REFERENCES public.person(id)
+    FOREIGN KEY (owned_by_person_id, workspace_id) REFERENCES public.person(id, workspace_id)
     ON UPDATE RESTRICT;
```

When I run Kanel, I get multiple warnings about a Circular reference:

```
Could not resolve circular reference {
  schemaName: 'public',
  tableName: 'workspace',
  columnName: 'id',
  onUpdate: 'CASCADE',
  onDelete: 'CASCADE',
  name: 'person:foreignKey(workspace_id,workspace.id)'
}
```

The generated Kysely types now include a union with `string`.

```diff
- workspaceId: ColumnType<WorkspaceId, WorkspaceId, WorkspaceId>;
+ workspaceId: ColumnType<WorkspaceId | string, WorkspaceId | string, WorkspaceId | string>;
```

And the Zod types are no longer branded.

```diff
- workspaceId: workspaceId,
+ workspaceId: z.string(),
```

I'm going to investigate why there is a circular reference error.

**Comment by kristiandupont** [comment]
Thank you, and sorry for dropping the ball on this!

**Comment by stayradiated** [comment]
No worries, thank you for creating and maintaining such a useful piece of software!

After a little `console.log` debugging, I think now understand what is happening here.
It's not a circular reference, but the same type is being resolved multiple times in a single pass.

Originally, the `document.workspace_id` column only has a single reference:

```mermaid
flowchart LR
    D1["document.workspace_id"] --> W1["workspace.id"]
```

After creating a composite foreign key, the `document.workspace_id` now has two references.
While resolving the type, the column `workspace.id` is resolved twice - which gets accidentally flagged as a circular reference, and `unknown` is returned on the second branch.

```mermaid
flowchart LR
    D1["document.workspace_id"] --> W1["workspace.id"]
    D1["document.workspace_id"] --> P1["person.workspace_id"]
    P1["person.workspace_id"] --> W1
```

One solution could be to replace the `visited` set with a map of the previously returned value.
This should handle both cases:
1. Detect circular dependencies and prevent the code from looping indefinitely
2. Handle cases like the above where the same column is visited multiple times while resolving a type

```typescript

// note, that in this scenario, the `getTypeFromReferences` function no longer checks for circular dependencies
// it should leave that to `resolveType` to handle

const resolveType = (c, d, config, cache = new Map()) => {
  if (cache.has(c)) {
    const { state, value } = cache.get(c);
    if (state === "pending") {
      console.warn("Circular reference detected", c);
      return "unknown";
    }
    return value;
  }
  cache.set(c, { state: "pending", value: undefined });

  const type = (() => {
       // existing logic for resolveType
       // ...
  })

  cache.set(c, { state: "resolved", value: type });
  return type;
};
```

I'm going to patch my version of `kanel` with this change locally and see if that works as expected while I continue re-factoring my database schema. 

Then I'll raise a PR to apply this patch, if you think that this is a good solution.

**Comment by stayradiated** [comment]
I've raised a PR here to fix this issue: https://github.com/kristiandupont/kanel/pull/642
